### DISCUSSION TITLE
Does a github action create a docker image or equivalent when it runs?

### DISCUSSION THREAD

**Discussion by anmol-aidora** [nektos/act#2499] [Q&A]
Hi 

If a Github Action image equivalent is actually created at the time of its first run, the next time it runs, it should ideally take less time to run, as the image equivalent has already been built. It will be just like spawning a container from a pre-built docker image when the action runs second time onwards.

This question is about the real Github Action and not about act. We know that act is based on Docker API, this question will answer itself and there wouldn't be any point asking it.

Thank you

**Comment by wellkilo** [comment]
GitHub Actions does not build one reusable “workflow image” for the entire workflow on its first run.

On GitHub-hosted runners, each job gets a fresh runner VM. GitHub starts from a prebuilt runner image, but then your job independently checks out code, downloads actions, pulls job/service container images, restores declared caches, and executes steps. The VM is discarded after the job, so arbitrary filesystem state from one run is not kept for the next run.

There are several separate caches that can make later runs faster:

- action/container artifacts may be cached by GitHub's infrastructure, but that is not a workflow-level contract;
- actions/cache restores paths under keys you define;
- Docker/BuildKit layer caching works only when you explicitly configure a cache backend (for example gha or registry) or use a persistent self-hosted builder;
- self-hosted runners may retain their machine/container cache, but workflows should not rely on undeclared leftovers.

A uses: docker://... step pulls an existing image. A local Docker action with Dockerfile may be built for the job, but that image is not automatically promoted to a durable cache for later hosted jobs.

So optimize repeat runs by declaring caches and publishing reusable images, not by assuming the first workflow run creates a reusable equivalent image. act differs because it uses your local Docker daemon, whose image/layer cache naturally persists between invocations; its README describes that local Docker execution model: [How Does It Work?](https://github.com/nektos/act/blob/4f411281417e88660bea1c1a1749aa71ae0bd60f/README.md#how-does-it-work).
